presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionResult.java

```diff
@@ -59,9 +64,27 @@ public HivePartitionResult(
         return partitionColumns;
     }
 
-    public List<HivePartition> getPartitions()
+    public Iterator<HivePartition> getPartitions()
     {
-        return partitions;
+        return partitions.iterator();
+    }
+
+    public List<HivePartition> getPartitionsAsList(int maxSize)
+    {
+        ImmutableList.Builder<HivePartition> partitionList = ImmutableList.builder();
+        int listSize = 0;
+        Iterator<HivePartition> iterator = partitions.iterator();
+        while (iterator.hasNext()) {
+            HivePartition partition = iterator.next();
+            partitionList.add(partition);
+            if (listSize++ == maxSize) {
```

Comment:
`++listSize > maxSize` ? (technically, this is the same, i know)
Or better yet, place `listSize++` before the `if`.
Or even, make `partitionList` an `ArrayList` and use `list.size()` (i know, this is old-school). You can still `ImmutableList.copyOf(partitionList)` and the end to maintain returning immutable list